src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala

```diff
@@ -353,35 +240,226 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+class QueryCompiler(cq : ConjunctiveQuery, ss: CompilationState) {
+  // This is responsible for schema elements within a given query, e.g.,
+  // what is the canonical version of x? (i.e., the first time it is
+  // mentioned in the body. This is useful to translate to SQL (join
+  // conditions, select, etc.)
+  var query_schema = new HashMap[ String, Tuple2[String,Variable] ]()
 
   // maps each variable name to a canonical version of itself (first occurence in body in left-to-right order)
   // index is the index of the subgoal/atom this variable is found in the body.
   // variable is the complete Variable type for the found variable.
   def generateCanonicalVar()  = {
-    q.bodies(0).zipWithIndex.foreach {
-      case (Atom(relName,terms),index) =>  {
-        terms.zipWithIndex.foreach { case (expr, i) =>
-          expr match {
-            case VarExpr(v) => 
-              if (! (query_schema contains v) )
-                query_schema += { v -> (index, Variable(v,relName,i) ) }
-            case _ =>
-          }
+    def generateCanonicalVarFromAtom(a: Atom, index: String) {
+      a.terms.zipWithIndex.foreach { case (expr, i) =>
+        expr match {
+          case VarExpr(v) =>
+            if (! (query_schema contains v) )
+              query_schema += { v -> (index, Variable(v,a.name,i) ) }
+          case _ =>
         }
       }
     }
+    // For quantified body, we need to recursively add variables from the atoms inside the quantified body
+    def generateCanonicalVarFromBodies(bodies: List[Body], indexPrefix: String) {
+      bodies.zipWithIndex.foreach {
+        case (a: Atom, index) => generateCanonicalVarFromAtom(a, s"${indexPrefix}${index}")
+        // we need to recursively handle the atoms inside the modifier
+        case (a: QuantifiedBody, index) => generateCanonicalVarFromBodies(a.bodies, s"${indexPrefix}${index}_")
+        case _ =>
+      }
+    }
+
+    generateCanonicalVarFromBodies(cq.bodies(0), "")
   }
   generateCanonicalVar() // initialize
 
   // accessors
-  def getBodyIndex( varName : String ) : Int = { query_schema(varName)._1 }
+  def getBodyIndex( varName : String )       = { query_schema(varName)._1 }
   def getVar(varName : String ) : Variable   = { query_schema(varName)._2 }
+
+  // compile a variable name to its canonical relation and column in R{index}.{name} style
+  def compileVariable(v: String) = {
+    val index = getBodyIndex(v)
+    val name  = ss.resolveName(getVar(v))
+    s"R${index}.${name}"
+  }
+
+  // compile alias
+  def compileAlias(e: Expr, alias: AliasStyle) = {
+    e match {
+      case VarExpr(v) => {
+        val index = getBodyIndex(v)
+        val variable = getVar(v)
+        val name  = ss.resolveName(variable)
+        val relation = variable.relName
+        alias match {
+          case NoAlias => ""
+          case HeadAlias => s""" AS "${relation}.R${index}.${name}" """
+          case WeightAlias => s"${relation}.R${index}.${name}"
```

Comment:
Overall, it became much clearer, but I'm still unhappy with this alias part.  I feel like we're almost there untying this ugly knot.  It seems there are several factors that unnecessarily complicate things here:
1. sometimes we want the alias to be prefixed with ` AS ` and sometimes not.
2. sometimes we don't want the alias at all.
3. only `VarExpr` case considers the `AliasStyle` which seems very odd.
4. each case seems to be quite independent from others, meaning we could split this into several smaller pieces.

Initially, it seemed `compileAlias` could be inlined into `compileExprWithAlias` below.  However, this is called from one extra place when deciding weight tying argument (a non-SQL part) with `WeightAlias`.  The other cases when a variable AliasStyle is passed is via `generateSQLHead`, which is `HeadAlias` only for one out of three cases.

How does the following sound?
1. `compileAlias` could be a simple function that maps to a deterministic alias for every expr, and 
2. `generateSQLHead` could append the aliases with ` AS ` if necessary.
3. Then `compileExprWithAlias` could be inlined and removed.
4. The `WeightAlias` becomes unnecessary as the caller can just get the alias from `compileAlias`.
5. `HeadAlias`/`NoAlias` becomes a boolean flag for `generateSQLHead`.